modules/repofiles/delete.go

```diff
@@ -0,0 +1,230 @@
+func DeleteRepoFile(repo *models.Repository, doer *models.User, opts *DeleteRepoFileOptions) (*api.FileResponse, error) {
+	if repo == nil {
+		return nil, fmt.Errorf("repo cannot be nil")
+	}
+	if doer == nil {
+		return nil, fmt.Errorf("doer cannot be nil")
+	}
+	if opts == nil {
+		return nil, fmt.Errorf("opts cannot be nil")
+	}
+
+	// If no branch name is set, assume the repo's default branch
+	if opts.OldBranch == "" {
+		opts.OldBranch = repo.DefaultBranch
+	}
+	if opts.NewBranch == "" {
+		opts.NewBranch = opts.OldBranch
+	}
+
+	// oldBranch must exist for this operation
+	if _, err := repo.GetBranch(opts.OldBranch); err != nil {
+		return nil, err
+	}
+
+	// A NewBranch can be specified for the file to be created/updated in a new branch.
+	// Check to make sure the branch does not already exist, otherwise we can't proceed.
+	// If we aren't branching to a new branch, make sure user can commit to the given branch
+	if opts.NewBranch != opts.OldBranch {
+		newBranch, err := repo.GetBranch(opts.NewBranch)
+		if git.IsErrNotExist(err) {
+			return nil, err
+		}
+		if newBranch != nil {
+			return nil, models.ErrBranchAlreadyExists{
+				BranchName: opts.NewBranch,
+			}
+		}
+	} else {
+		if protected, _ := repo.IsProtectedBranchForPush(opts.OldBranch, doer); protected {
+			return nil, models.ErrUserCannotCommit{
+				UserName: doer.LowerName,
+			}
+		}
+	}
+
+	// Check that the path given in opts.treeName is valid (not a git path)
+	treePath := CleanUploadFileName(opts.TreePath)
+	if treePath == "" {
+		return nil, models.ErrFilenameInvalid{
+			Path: opts.TreePath,
+		}
+	}
+
+	message := strings.TrimSpace(opts.Message)
+
+	author, committer := GetAuthorAndCommitterUsers(opts.Committer, opts.Author, doer)
+
+	t, err := NewTemporaryUploadRepository(repo)
+	defer t.Close()
+	if err != nil {
+		return nil, err
+	}
+	if err := t.Clone(opts.OldBranch); err != nil {
+		return nil, err
+	}
+	if err := t.SetDefaultIndex(); err != nil {
+		return nil, err
+	}
+
+	if opts.LastCommitID == "" {
+		commitID, err := t.GetLastCommit()
+		if err != nil {
+			return nil, err
+		}
+		opts.LastCommitID = commitID
+	}
+
+	gitRepo, err := git.OpenRepository(repo.RepoPath())
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the commit of the original branch
+	commit, err := gitRepo.GetBranchCommit(opts.OldBranch)
+	if err != nil {
+		return nil, err // Couldn't get a commit for the branch
+	}
+
+	// Get the files in the index
+	filesInIndex, err := t.LsFiles(opts.TreePath)
+	if err != nil {
+		return nil, fmt.Errorf("DeleteRepoFile: %v", err)
+	}
+
+	// Find the file we want to delete in the index
+	inFilelist := false
+	for _, file := range filesInIndex {
+		if file == opts.TreePath {
+			inFilelist = true
+		}
+	}
+	if !inFilelist {
+		return nil, models.ErrRepoFileDoesNotExist{
+			Path: opts.TreePath,
+		}
+	}
+
+	// Get the entry of treePath and check if the SHA given is the same as the file
+	entry, err := commit.GetTreeEntryByPath(treePath)
```

Comment:
OK, I created the TemporaryUploadRepository structure to avoid using the relatively bulky and probably incorrect GitRepository things and to be as minimal as possible. (For example none of those things use `--` to stop argument processing, nor do they use `-z` to zero pad and there is a lack of porcelain options in general.)

If we're going to use the GitRepository stuff the work with the LsFiles is unnecessary.

Either, use the GitRepository or improve the LsFiles code to give out the SHA code. (`git ls-files -z --stage -- FILE_NAME`)

Reply:
Will look into this. Thanks.

Reply:
Refactored for the TempRepo to have a pointer to a gitRepo so it can be used with the TempRepo's path to the cloned repo